apps/system/js/advanced_telemetry.js

```diff
@@ -0,0 +1,426 @@
+  // long do we wait before trying it again?
+  // Can be overridden with metrics.advancedtelemetry.retryTimeout setting.
+  AT.RETRY_INTERVAL = 60 * 60 * 1000;        // 1 hour
+
+  // Telemetry payload version
+  AT.TELEMETRY_VERSION = 4;
+
+  // App name (static for Telemetry)
+  AT.TELEMETRY_APP_NAME = 'FirefoxOS';
+
+
+  AT.prototype.start = function start() {
+    loginfo('Starting AdvancedTelemetry');
+    this.advancedTelemetryEnabledListener =
+      function advancedTelemetryEnabledListener(enabled) {
+        if (enabled) {
+          loginfo('Start Collecting');
+          this.startCollecting();
+        }
+        else {
+          loginfo('Stop Collecting');
+          this.stopCollecting();
+        }
+      }.bind(this);
+
+    SettingsListener.observe(AT.TELEMETRY_ENABLED_KEY,
+      false, this.advancedTelemetryEnabledListener);
+  };
+
+  AT.prototype.startCollecting = function startCollecting(done) {
+    var self = this;
+    // If we're already running there is nothing to start
+    if (this.collecting) {
+      return;
+    }
+    this.collecting = true;
+
+    this.metrics = new BatchTiming(true);
+
+    getDeviceID();
+
+    function getDeviceID() {
+      asyncStorage.getItem(DEVICE_ID_KEY, function(value) {
+        if (value) {
+          self.deviceID = value;
+        }
+        else {
+          // Our device id does not need to be unique, just probably unique.
+          self.deviceID = uuid();
+          asyncStorage.setItem(DEVICE_ID_KEY, self.deviceID);
+        }
+
+        // Move on to the next step in the startup process
+        registerHandlers();
+      });
+    }
+
+    function registerHandlers() {
+      // Basic event handlers
+      EVENT_TYPES.forEach(function (type) {
+        window.addEventListener(type, self);
+      });
+
+      self.startBatch();
+
+      if (done) {
+        loginfo('Initialization Complete');
+        done();
+      }
+    }
+  };
+
+  // This method shuts everything down and is only exposed for unit testing.
+  // Note that this is not the same as the stopCollecting() method which is
+  // used to stop data collection but keep the module running.
+  AT.prototype.stop = function stop() {
+
+  };
+
+  // Reset (or initialize) the AdvancedTelemetry instance variables
+  AT.prototype.reset = function() {
+    this.collecting = false;
+    this.metrics = null;
+    this.deviceID = null;
+  };
+
+  AT.prototype.handleEvent = function handleEvent(e) {
+    switch (e.type) {
+      case HUDMETRICS:
+        this.transmit(e.detail);
+        break;
+      default:
+        break;
+    }
+  };
+
+  // Signal gecko to send the histograms with a telemetry message.
+  AT.prototype.getPayload = function getPayload() {
+    console.info('telemetry|MGMT|TIMETOSHIP');
+  };
+
+  // Start a timer to notify when to send the payload.
+  AT.prototype.startBatch = function startBatch() {
+    var self = this;
+    this.timer = setTimeout(function() {
```

Comment:
is `self.getPayload()` supposed to be called continuously on an interval or just once? you use `this.interval` as the timeout value, but `self.getPayload()` will only be called once here because you're using `setTimeout()` and not `setInterval()`